metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/execution/WorkflowPostProcessor.java

```diff
@@ -50,49 +66,91 @@ public WorkflowPostProcessor(DepublishRecordIdDao depublishRecordIdDao,
+          throws DpsException {
 
     final PluginType pluginType = plugin.getPluginType();
     LOGGER.info("Starting postprocessing of plugin {} in dataset {}.", pluginType, datasetId);
     //Reset depublish status if index to PUBLISH
     if (pluginType == PluginType.PUBLISH) {
       publishPostProcess(datasetId);
     } else if (pluginType == PluginType.DEPUBLISH) {
-      depublishPostProcess(plugin, datasetId);
+      depublishPostProcess((DepublishPlugin) plugin, datasetId);
     }
     LOGGER.info("Finished postprocessing of plugin {} in dataset {}.", pluginType, datasetId);
   }
 
   private void publishPostProcess(String datasetId) {
-    depublishRecordIdDao.markRecordIdsWithDepublicationStatus(datasetId,
+    depublishRecordIdDao.markRecordIdsWithDepublicationStatus(datasetId, null,
         DepublicationStatus.PENDING_DEPUBLICATION, null);
   }
 
-  private void depublishPostProcess(AbstractExecutablePlugin<?> plugin, String datasetId) {
-    final boolean datasetDepublish = ((DepublishPluginMetadata) plugin.getPluginMetadata())
-        .isDatasetDepublish();
-    if (datasetDepublish) { //Reset depublish status of records if depublishing dataset
-      depublishRecordIdDao.markRecordIdsWithDepublicationStatus(datasetId,
-          DepublicationStatus.PENDING_DEPUBLICATION, null);
-      //Find latest PUBLISH Type Plugin and set dataStatus
-      final PluginWithExecutionId<MetisPlugin> latestSuccessfulPlugin = workflowExecutionDao
-          .getLatestSuccessfulPlugin(datasetId, OrchestratorService.PUBLISH_TYPES);
-      if (Objects.nonNull(latestSuccessfulPlugin) && Objects
-          .nonNull(latestSuccessfulPlugin.getPlugin())) {
-        final WorkflowExecution workflowExecutionToUpdate = workflowExecutionDao
-            .getById(latestSuccessfulPlugin.getExecutionId());
-        final Optional<AbstractMetisPlugin> metisPluginWithType = workflowExecutionToUpdate
-            .getMetisPluginWithType(latestSuccessfulPlugin.getPlugin().getPluginType());
-        if (metisPluginWithType.isPresent()) {
-          metisPluginWithType.get().setDataStatus(DataStatus.DELETED);
-          workflowExecutionDao.updateWorkflowPlugins(workflowExecutionToUpdate);
-        }
+  private void depublishPostProcess(DepublishPlugin plugin, String datasetId) throws DpsException {
+    if (plugin.getPluginMetadata().isDatasetDepublish()) {
+      depublishDatasetPostProcess(datasetId);
+    } else {
+      depublishRecordPostProcess(plugin, datasetId);
+    }
+  }
+
+  private void depublishDatasetPostProcess(String datasetId){
+
+    // Set all depublished records back to PENDING.
+    depublishRecordIdDao.markRecordIdsWithDepublicationStatus(datasetId, null,
+            DepublicationStatus.PENDING_DEPUBLICATION, null);
+
+    // Find latest PUBLISH Type Plugin and set dataStatus to DELETED.
+    final PluginWithExecutionId<MetisPlugin> latestSuccessfulPlugin = workflowExecutionDao
```

Comment:
IDE complaints about generic type

Reply:
I think this is outside scope... it's just what the called method returns. I think we need a separate tech debt ticket to solve those issues.

Reply:
Okay no problem